File: lab06/py/Phi.py
from math import isqrt
from collections import defaultdict
from typing import List, Dict, Tuple

def phi(n: int) -> int:
    current_factorization: Dict[int, int] = factorization(n)
    
    result = 1
    
    for prime, exponent in current_factorization.items():
        result *= (prime - 1)
        result *= prime ** (exponent - 1)
    
    return result

def factorization(n: int) -> Dict[int, int]:
    factors = factor(n)

    grouped_factors = defaultdict(int)
    for currentFactor in factors:
        grouped_factors[currentFactor] += 1

    return dict(grouped_factors)

def factor(n: int) -> List[int]:
    result = []

    # Remove powers of 2, 3, 5
    for prime in [2, 3, 5]:
        n = remove_factors(n, result, prime)

    d = 0

    # Trial division with selected forms of d
    square_root = isqrt(n)
    while n > 1 and d <= square_root:
        forms_of_d = [1, 7, 11, 13, 17, 19, 23, 29]

        for current_residue in forms_of_d:
            current_divisor = d + current_residue

            if current_divisor != 1 and n % current_divisor == 0:
                n = remove_factors(n, result, current_divisor)

        d += 30  # Move to the next set of forms

    if n > 1:
        result.append(n)

    return result

def remove_factors(n: int, factors: List[int], prime: int) -> int:
    while n % prime == 0:
        factors.append(prime)
        n //= prime

    return n

def main():
    n = int(input("Enter number n: "))
    result = phi(n)
    print(f"ϕ({n}) = {result}")

main()

File: lab06/cpp/Phi.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>

using namespace std;

map<int, int> factorization(int n);
vector<int> factor(int n);
int remove_factors(int &n, vector<int> &factors, int prime);


map<int, int> factorization(int n) {
    vector<int> factors = factor(n);
    map<int, int> grouped_factors;
    for (int f : factors) {
        grouped_factors[f]++;
    }
    return grouped_factors;
}

vector<int> factor(int n) {
    vector<int> result;
    for (int prime : {2, 3, 5}) {
        remove_factors(n, result, prime);
    }
    int d = 0;
    int square_root = sqrt(n);
    while (n > 1 && d <= square_root) {
        for (int current_residue : {1, 7, 11, 13, 17, 19, 23, 29}) {
            int current_divisor = d + current_residue;
            if (current_divisor != 1 && n % current_divisor == 0) {
                remove_factors(n, result, current_divisor);
            }
        }
        d += 30;
    }
    if (n > 1) {
        result.push_back(n);
    }
    return result;
}

int remove_factors(int &n, vector<int> &factors, int prime) {
    while (n % prime == 0) {
        factors.push_back(prime);
        n /= prime;
    }
    return n;
}

int phi(int n) {
    map<int, int> current_factorization = factorization(n);
    int result = 1;
    
    for (const auto &[prime, exponent] : current_factorization) {
        result *= (prime - 1);
        result *= pow(prime, exponent - 1);
    }
    
    return result;
}

int main() {
    int n;
    cout << "Enter number n: ";
    cin >> n;
    int result = phi(n);
    cout << "ϕ(" << n << ") = " << result << endl;
    return 0;
}
